phpBB/phpbb/auth/provider/oauth/oauth.php

```diff
@@ -654,114 +696,172 @@ protected function link_account_perform_link(array $data)
 		}
 
 		// Link account
-		$sql = 'INSERT INTO ' . $this->auth_provider_oauth_token_account_assoc . '
-			' . $this->db->sql_build_array('INSERT', $data);
+		$sql = 'INSERT INTO ' . $this->oauth_account_table . ' ' . $this->db->sql_build_array('INSERT', $data);
 		$this->db->sql_query($sql);
 
 		/**
 		 * Event is triggered after user links account.
 		 *
 		 * @event core.auth_oauth_link_after
-		 * @var    array    data    User row
+		 * @var array	data	User row
 		 * @since 3.1.11-RC1
 		 */
-		$vars = array(
+		$vars = [
 			'data',
-		);
+		];
 		extract($this->dispatcher->trigger_event('core.auth_oauth_link_after', compact($vars)));
 	}
 
 	/**
-	* {@inheritdoc}
-	*/
-	public function logout($data, $new_session)
+	 * Returns a new service object.
+	 *
+	 * @param string			$provider		The name of the provider
+	 * @param token_storage		$storage		Token storage object
+	 * @param string			$query			The query string used for the redirect uri
+	 * @return ServiceInterface
+	 * @throws exception						When OAuth service was not created
+	 */
+	protected function get_service($provider, token_storage $storage, $query)
 	{
-		// Clear all tokens belonging to the user
-		$storage = new \phpbb\auth\provider\oauth\token_storage($this->db, $this->user, $this->auth_provider_oauth_token_storage_table, $this->auth_provider_oauth_state_table);
-		$storage->clearAllTokens();
+		$service_name = $this->get_service_name($provider);
 
-		return;
-	}
+		/** @see \phpbb\auth\provider\oauth\service\service_interface::get_service_credentials */
+		$service_credentials = $this->service_providers[$service_name]->get_service_credentials();
```

Comment:
Isn't any method to get services by name?, offsetGet maybe?

Reply:
As the service collection implements ArrayObject this type of call will result in offsetGet() being called.